doc/design.md

```diff
@@ -0,0 +1,922 @@
+
+All relevant high-level classes and functions are fully exposed via a Python interface.
+- The C++ API is modern and is thus easily mapped to a "pythonic" dataset library.
+- `Dataset` itself is very similar to Python `dict`.
+- Data of variables containing plain old data such as floating point values is exposed as readable and writable `numpy.ndarray`.
+  These are true multi-dimensional arrays, i.e., in contrast to the current `API::MatrixWorkspace::readX()` and `API::MatrixWorkspace::dataX()` it is *not* a separate array for every spectrum.
+
+
+## <a name="design-components"></a>Design components
+
+The following is only an incomplete description.
+For more details and examples the prototype code serves as a reference.
+A good starting point that gives an overview of how `Dataset` may be *used* is given by [`TableWorkspace` example](../test/TableWorkspace_test.cpp), [`Workspace2D` example](../test/Workspace2D_test.cpp), and [`EventWorkspace` example](../test/EventWorkspace_test.cpp).
+
+
+### <a name="design-components-dataset"></a>`class Dataset`
+
+`class Dataset` is literally just a set or list of variables.
+- Variables can be inserted and removed.
+- Iteration over variables is possible.
+- Variables are identified by a combination of their tag and (for data variables) their name.
+  Duplicate tag/name combinations are forbidden, i.e., `Dataset` behaves like a set.
+
+The current implementation also contains a `Dimensions` objects, holding the space spanned by the combination of all the variables dimensions.
+However, this is not necessary and may change in the actual implementation.
+
+There are very few methods for direct manipulation of `Dataset`:
+- `setSlice` to set a slice of the dataset.
+
+The current implementation in the prototype is mostly complete regarding the actual `Dataset` interface.
+Most other functionality would probably be added as free functions.
+- Some interface cleanup for finding variables or checking for existence of certain tags/names is required.
+- `merge` is currently a member method, but should probably be made a free function.
+  Given the copy-on-write mechanism there is no advantage in having it as a method.
+- Insertion mechanism for bin-edge variables needs to be clarified.
+
+
+### <a name="design-components-variable"></a>`class Variable`
+
+`class Variable` can hold a N-dimensional array of arbitrary data, with some additions that make it self-describing:
+- Tag and name identifying the variable.
+- Unit of the data held by the variable.
+- Dimensions of the data held by the variable, i.e., the side lengths of a multi-dimensional cube of data, using `class Dimensions`.
+- Data, stored in an array-like container.
+
+In the current implementation, as much as possible has been moved out of the header to reduce compile times due to templated code.
+The interface implemented currently should be mostly complete with little change required for the final implementation.
+
+##### Type erasure
```

Comment:
I feel that `Type Erasure` is significantly important to the overall design and should be introduced/mentioned earlier.

Reply:
It is here: https://github.com/mantidproject/dataset/pull/1/files/8da5efdffe574ec9eb4897a6d65c7089354e5393#diff-1ab4257b823a2299454fd7d83cfa11ebR197.

Reply:
Ok. That's fine then.